File: src/gateway/routing/guards.py
"""Post-processor guard checks.

These are hard rejection rules that run AFTER the Processor LLM
has made its decision. The LLM decides trade validity using
TA + Macro + RAG context. Guards are the final safety net that
enforce non-negotiable operational rules before execution.

Guards do NOT decide trade validity - the processor does.
Guards BLOCK execution when hard safety rules are violated.
"""

from __future__ import annotations

import time
from datetime import UTC, datetime
from typing import Optional

from engine.shared.logging import get_logger
from engine.shared.models.events import NEWS_LOCKOUT_MINUTES
from gateway.constants import GuardRule, GuardVerdict
from gateway.context.models import (
    GuardCheckResult,
    GuardEvaluationResult,
    MacroResult,
    ProcessorOutput,
    TASymbolResult,
)
from gateway.observability.metrics import GATEWAY_GUARD_DURATION, GATEWAY_GUARD_REJECTIONS

logger = get_logger(__name__)


class GuardEvaluator:
    """Evaluates all post-processor guard rules.

    Each guard is a pure check that returns PASS, REJECT, or WARN.
    The evaluator aggregates all results and determines the overall verdict.
    """

    def evaluate(
        self,
        *,
        processor_output: ProcessorOutput,
        ta_result: TASymbolResult,
        macro_result: MacroResult,
        trace_id: Optional[str] = None,
    ) -> GuardEvaluationResult:
        """Run all guard checks and return aggregated result."""
        start = time.monotonic()

        checks: list[GuardCheckResult] = [
            self._check_news_proximity(macro_result),
            self._check_session_restriction(ta_result),
            self._check_counter_trend(processor_output, ta_result),
            self._check_weekend_gap_risk(),
            self._check_low_liquidity_hours(),
        ]

        blocking: list[str] = []
        overall = GuardVerdict.PASS

        for check in checks:
            if check.verdict == GuardVerdict.REJECT:
                overall = GuardVerdict.REJECT
                blocking.append(check.rule.value)
                GATEWAY_GUARD_REJECTIONS.labels(rule=check.rule.value).inc()
            elif check.verdict == GuardVerdict.WARN and overall != GuardVerdict.REJECT:
                overall = GuardVerdict.WARN

        elapsed = time.monotonic() - start
        GATEWAY_GUARD_DURATION.observe(elapsed)

        result = GuardEvaluationResult(
            checks=checks,
            overall_verdict=overall,
            blocking_rules=blocking,
        )

        logger.info(
            "guard_evaluation_completed",
            extra={
                "overall_verdict": overall.value,
                "blocking_rules": blocking,
                "checks_total": len(checks),
                "duration_ms": round(elapsed * 1000, 1),
                "trace_id": trace_id,
            },
        )

        return result

    @staticmethod
    def _check_news_proximity(macro: MacroResult) -> GuardCheckResult:
        """MR-REJECT-001: No entries within NEWS_LOCKOUT_MINUTES of high-impact news."""
        calendar = macro.calendar
        if not calendar:
            return GuardCheckResult(
                rule=GuardRule.NEWS_PROXIMITY,
                verdict=GuardVerdict.PASS,
                reason="No calendar data available",
            )

        events = calendar.get("events", [])
        now = datetime.now(UTC)

        for event in events:
            impact = event.get("impact", "").upper()
            if impact != "HIGH":
                continue

            event_time_str = event.get("event_time")
            if not event_time_str:
                continue

            try:
                if isinstance(event_time_str, str):
                    event_time = datetime.fromisoformat(event_time_str.replace("Z", "+00:00"))
                else:
                    event_time = event_time_str

                if event_time.tzinfo is None:
                    event_time = event_time.replace(tzinfo=UTC)

                minutes_until = (event_time - now).total_seconds() / 60

                if 0 <= minutes_until <= NEWS_LOCKOUT_MINUTES:
                    return GuardCheckResult(
                        rule=GuardRule.NEWS_PROXIMITY,
                        verdict=GuardVerdict.REJECT,
                        reason=(
                            f"High-impact event '{event.get('event_name', 'unknown')}' "
                            f"in {int(minutes_until)} minutes (lockout: {NEWS_LOCKOUT_MINUTES}min)"
                        ),
                        metadata={"event_name": event.get("event_name"), "minutes_until": minutes_until},
                    )
            except (ValueError, TypeError):
                continue

        return GuardCheckResult(
            rule=GuardRule.NEWS_PROXIMITY,
            verdict=GuardVerdict.PASS,
            reason="No high-impact events within lockout window",
        )

    @staticmethod
    def _check_session_restriction(ta: TASymbolResult) -> GuardCheckResult:
        """MR-REJECT-002: No entries during restricted sessions (Asian for most pairs)."""
        now = datetime.now(UTC)
        hour = now.hour

        is_asian = 0 <= hour < 7

        if not is_asian:
            return GuardCheckResult(
                rule=GuardRule.SESSION_RESTRICTION,
                verdict=GuardVerdict.PASS,
                reason=f"Current hour {hour} UTC is outside Asian session",
            )

        symbol = ta.symbol.upper()
        is_jpy_pair = "JPY" in symbol
        is_aud_pair = "AUD" in symbol
        is_nzd_pair = "NZD" in symbol

        if is_jpy_pair or is_aud_pair or is_nzd_pair:
            return GuardCheckResult(
                rule=GuardRule.SESSION_RESTRICTION,
                verdict=GuardVerdict.PASS,
                reason=f"{symbol} is active during Asian session",
            )

        return GuardCheckResult(
            rule=GuardRule.SESSION_RESTRICTION,
            verdict=GuardVerdict.REJECT,
            reason=f"Asian session restriction: {symbol} should not be traded 00:00-07:00 UTC",
            metadata={"hour_utc": hour, "symbol": symbol},
        )

    @staticmethod
    def _check_counter_trend(
        processor: ProcessorOutput,
        ta: TASymbolResult,
    ) -> GuardCheckResult:
        """MR-REJECT-006: Counter-trend without HTF CHoCH = NO SETUP.

        If the processor approved a trade that goes against the HTF trend
        and there is no HTF CHoCH in the snapshot, reject it.
        """
        if not processor.trade_valid:
            return GuardCheckResult(
                rule=GuardRule.COUNTER_TREND_NO_CHOCH,
                verdict=GuardVerdict.PASS,
                reason="Trade not valid, guard not applicable",
            )

        snapshot = ta.snapshot or {}
        trend = snapshot.get("trend_direction", "NEUTRAL")
        direction = (processor.direction or "").upper()

        is_counter = (
            (trend == "BULLISH" and direction in ("SHORT", "BEARISH", "SELL"))
            or (trend == "BEARISH" and direction in ("LONG", "BULLISH", "BUY"))
        )

        if not is_counter:
            return GuardCheckResult(
                rule=GuardRule.COUNTER_TREND_NO_CHOCH,
                verdict=GuardVerdict.PASS,
                reason="Trade aligns with HTF trend",
            )

        choch_events = snapshot.get("choch_events", {})
        choch_count = choch_events.get("count", 0) if isinstance(choch_events, dict) else 0

        if choch_count > 0:
            return GuardCheckResult(
                rule=GuardRule.COUNTER_TREND_NO_CHOCH,
                verdict=GuardVerdict.WARN,
                reason="Counter-trend trade with CHoCH detected - proceed with caution",
                metadata={"choch_count": choch_count},
            )

        htf = ta.htf_timeframe or "HTF"
        return GuardCheckResult(
            rule=GuardRule.COUNTER_TREND_NO_CHOCH,
            verdict=GuardVerdict.REJECT,
            reason=f"Counter-trend trade without {htf} CHoCH - rejected per MR-REJECT-006",
            metadata={"trend": trend, "direction": direction},
        )

    @staticmethod
    def _check_weekend_gap_risk() -> GuardCheckResult:
        """MR-REJECT-008: No new entries close to market close on Friday."""
        now = datetime.now(UTC)

        if now.weekday() == 4 and now.hour >= 20:
            return GuardCheckResult(
                rule=GuardRule.WEEKEND_GAP_RISK,
                verdict=GuardVerdict.REJECT,
                reason="Friday after 20:00 UTC - weekend gap risk",
                metadata={"day": "Friday", "hour": now.hour},
            )

        if now.weekday() in (5, 6):
            return GuardCheckResult(
                rule=GuardRule.WEEKEND_GAP_RISK,
                verdict=GuardVerdict.REJECT,
                reason="Weekend - market closed",
                metadata={"day": now.strftime("%A")},
            )

        return GuardCheckResult(
            rule=GuardRule.WEEKEND_GAP_RISK,
            verdict=GuardVerdict.PASS,
            reason="Not in weekend gap risk window",
        )

    @staticmethod
    def _check_low_liquidity_hours() -> GuardCheckResult:
        """MR-REJECT-009: Warn during known low-liquidity hours."""
        now = datetime.now(UTC)
        hour = now.hour

        if 21 <= hour or hour < 1:
            return GuardCheckResult(
                rule=GuardRule.LOW_LIQUIDITY_HOURS,
                verdict=GuardVerdict.WARN,
                reason=f"Low liquidity period: {hour}:00 UTC",
                metadata={"hour_utc": hour},
            )

        return GuardCheckResult(
            rule=GuardRule.LOW_LIQUIDITY_HOURS,
            verdict=GuardVerdict.PASS,
            reason=f"Normal liquidity hours: {hour}:00 UTC",
        )

File: src/gateway/internal/routing/guards.go
package routing

import (
	"fmt"
	"strings"
	"time"

	"github.com/rs/zerolog"

	"github.com/flamegreat-1/etradie/src/gateway/internal/constants"
	"github.com/flamegreat-1/etradie/src/gateway/internal/models"
	"github.com/flamegreat-1/etradie/src/gateway/internal/observability"
	"github.com/flamegreat-1/etradie/src/gateway/internal/querybuilder"
)

// GuardEvaluator evaluates all post-processor guard rules.
//
// Guards split into two phases by data dependency:
//
//   - Pre-LLM (deterministic): MR-REJECT-001, -002, -008, -009.
//     These depend only on time, symbol, and macro calendar data.
//     Running them before the processor LLM lets the cycle
//     short-circuit on a guaranteed rejection without spending
//     ~168k input tokens on a call whose result will be discarded.
//
//   - Post-LLM (LLM-dependent): MR-REJECT-006 (counter-trend).
//     Needs ProcessorOutput.Direction, so it can only run after
//     the processor LLM returns.
//
// Each guard still runs exactly once per cycle; the split is a
// scheduling change, not a duplication.
type GuardEvaluator struct {
	log zerolog.Logger
}

// NewGuardEvaluator creates a GuardEvaluator.
func NewGuardEvaluator() *GuardEvaluator {
	return &GuardEvaluator{log: observability.Logger("guard_evaluator")}
}

// EvaluatePreLLM runs the 4 deterministic guards that do not need
// the processor LLM output. The returned result carries exactly 4
// checks in the canonical order (001, 002, 008, 009). Callers that
// want all 5 checks in the final result must merge this with the
// post-LLM result via MergeResults.
//
// This is the cycle's first hard gate: when the result rejects,
// processSymbol skips the LLM call entirely.
func (g *GuardEvaluator) EvaluatePreLLM(
	taResult *models.TASymbolResult,
	macroResult *models.MacroResult,
	traceID string,
) *models.GuardEvaluationResult {
	start := time.Now()

	checks := []models.GuardCheckResult{
		checkHighImpactEventProximity(taResult, macroResult),
		checkSessionRestriction(taResult),
		checkWeekendGapRisk(taResult),
		checkLowLiquidityHours(taResult),
	}

	result := aggregate(checks)

	elapsed := time.Since(start).Seconds()
	observability.GatewayGuardDuration.Observe(elapsed)

	g.log.Info().
		Str("phase", "pre_llm").
		Str("overall_verdict", string(result.OverallVerdict)).
		Strs("blocking_rules", result.BlockingRules).
		Int("checks_total", len(checks)).
		Float64("duration_ms", elapsed*1000).
		Str("trace_id", traceID).
		Msg("guard_evaluation_completed")

	return result
}

// EvaluatePostLLM runs only the LLM-dependent guards. Today that is
// MR-REJECT-006 (counter-trend) which needs ProcessorOutput.Direction.
// The returned result carries exactly 1 check.
func (g *GuardEvaluator) EvaluatePostLLM(
	processorOutput *models.ProcessorOutput,
	taResult *models.TASymbolResult,
	traceID string,
) *models.GuardEvaluationResult {
	start := time.Now()

	checks := []models.GuardCheckResult{
		checkCounterTrend(processorOutput, taResult),
	}

	result := aggregate(checks)

	elapsed := time.Since(start).Seconds()
	observability.GatewayGuardDuration.Observe(elapsed)

	g.log.Info().
		Str("phase", "post_llm").
		Str("overall_verdict", string(result.OverallVerdict)).
		Strs("blocking_rules", result.BlockingRules).
		Int("checks_total", len(checks)).
		Float64("duration_ms", elapsed*1000).
		Str("trace_id", traceID).
		Msg("guard_evaluation_completed")

	return result
}

// MergeResults combines a pre-LLM and a post-LLM result into the
// single GuardEvaluationResult the dashboard and audit log expect.
// Checks are emitted in the canonical order (001, 002, 006, 008,
// 009) regardless of the order the phases ran in, so consumers do
// not need to know about the split.
//
// Either argument may be nil: when the cycle short-circuits pre-LLM
// there is no post-LLM result, and EvaluatePreLLM still returns the
// full set of pre-LLM checks.
func MergeResults(pre, post *models.GuardEvaluationResult) *models.GuardEvaluationResult {
	if pre == nil && post == nil {
		return &models.GuardEvaluationResult{OverallVerdict: constants.VerdictPass}
	}

	byRule := make(map[constants.GuardRule]models.GuardCheckResult, 5)
	if pre != nil {
		for _, c := range pre.Checks {
			byRule[c.Rule] = c
		}
	}
	if post != nil {
		for _, c := range post.Checks {
			byRule[c.Rule] = c
		}
	}

	canonical := []constants.GuardRule{
		constants.RuleHighImpactEventProximity,
		constants.RuleSessionRestriction,
		constants.RuleCounterTrendNoChoch,
		constants.RuleWeekendGapRisk,
		constants.RuleLowLiquidityHours,
	}

	ordered := make([]models.GuardCheckResult, 0, len(canonical))
	for _, rule := range canonical {
		if c, ok := byRule[rule]; ok {
			ordered = append(ordered, c)
		}
	}

	return aggregateNoMetrics(ordered)
}

// Evaluate runs ALL guard checks in a single call (legacy API).
//
// Kept so any caller that does not need the pre/post split can still
// produce a complete GuardEvaluationResult in one step. New call
// sites should prefer EvaluatePreLLM + EvaluatePostLLM + MergeResults
// so the deterministic checks can short-circuit before the LLM call.
func (g *GuardEvaluator) Evaluate(
	processorOutput *models.ProcessorOutput,
	taResult *models.TASymbolResult,
	macroResult *models.MacroResult,
	traceID string,
) *models.GuardEvaluationResult {
	pre := g.EvaluatePreLLM(taResult, macroResult, traceID)
	post := g.EvaluatePostLLM(processorOutput, taResult, traceID)
	return MergeResults(pre, post)
}

// aggregate computes the OverallVerdict and BlockingRules slice for a
// set of check results AND bumps the per-rule Prometheus rejection
// counter. Used by EvaluatePreLLM and EvaluatePostLLM where each check
// runs for the first (and only) time in a cycle.
func aggregate(checks []models.GuardCheckResult) *models.GuardEvaluationResult {
	var blocking []string
	overall := constants.VerdictPass

	for _, check := range checks {
		if check.Verdict == constants.VerdictReject {
			overall = constants.VerdictReject
			blocking = append(blocking, string(check.Rule))
			observability.GatewayGuardRejections.WithLabelValues(string(check.Rule)).Inc()
		} else if check.Verdict == constants.VerdictWarn && overall != constants.VerdictReject {
			overall = constants.VerdictWarn
		}
	}

	return &models.GuardEvaluationResult{
		Checks:         checks,
		OverallVerdict: overall,
		BlockingRules:  blocking,
	}
}

// aggregateNoMetrics computes OverallVerdict and BlockingRules WITHOUT
// touching Prometheus counters. Used by MergeResults to re-aggregate
// already-evaluated checks; the original aggregate call that produced
// each pre/post result has already incremented the per-rule counter
// exactly once, so re-counting here would double-count.
func aggregateNoMetrics(checks []models.GuardCheckResult) *models.GuardEvaluationResult {
	var blocking []string
	overall := constants.VerdictPass

	for _, check := range checks {
		if check.Verdict == constants.VerdictReject {
			overall = constants.VerdictReject
			blocking = append(blocking, string(check.Rule))
		} else if check.Verdict == constants.VerdictWarn && overall != constants.VerdictReject {
			overall = constants.VerdictWarn
		}
	}

	return &models.GuardEvaluationResult{
		Checks:         checks,
		OverallVerdict: overall,
		BlockingRules:  blocking,
	}
}

// MR-REJECT-001: No entries within the news lockout window of a
// high-impact economic-calendar event (NFP, CPI, PPI, FED rate
// decision, etc.) that affects one of the symbol's currencies.
//
// This is the DECISION-TIME gate. Because the trade's activation can
// be delayed (a LIMIT order rests at the broker for its TTL; an
// INSTANT watcher polls until price enters the zone), the same
// currency-scoped evaluation is re-run at fire time
// (RunConfirmationPulseWithParams) and at LIMIT placement / TTL time
// (CheckNewsWindow RPC). Here the trading style is not yet known
// (pre-LLM), so the normal lockout window is used; the wider
// scalping window is enforced later where the style is known.
func checkHighImpactEventProximity(ta *models.TASymbolResult, macro *models.MacroResult) models.GuardCheckResult {
	// 24/7 markets have no fiat calendar exposure; the orchestrator
	// also passes a nil macro for them. EvaluateNewsWindow returns
	// "no exposure" for such symbols, so they pass without needing a
	// calendar.
	if Is247Market(ta.Symbol) {
		return models.GuardCheckResult{
			Rule: constants.RuleHighImpactEventProximity, Verdict: constants.VerdictPass,
			Reason: "News proximity does not apply to 24/7 markets",
		}
	}

	var calendar map[string]interface{}
	if macro != nil {
		calendar = macro.Calendar
	}

	status := EvaluateNewsWindow(calendar, ta.Symbol, time.Now().UTC(), constants.HighImpactEventLockoutMinutes)

	if !status.DataAvailable {
		// Fail closed (N3): a non-24/7 symbol with no calendar data
		// must not trade blind into a possible high-impact event.
		return models.GuardCheckResult{
			Rule:     constants.RuleHighImpactEventProximity,
			Verdict:  constants.VerdictReject,
			Reason:   status.Reason,
			Metadata: map[string]interface{}{"data_available": false},
		}
	}

	if status.Locked {
		return models.GuardCheckResult{
			Rule:    constants.RuleHighImpactEventProximity,
			Verdict: constants.VerdictReject,
			Reason:  status.Reason,
			Metadata: map[string]interface{}{
				"event_name":    status.EventName,
				"currency":      status.Currency,
				"minutes_until": status.MinutesUntil,
			},
		}
	}

	return models.GuardCheckResult{
		Rule: constants.RuleHighImpactEventProximity, Verdict: constants.VerdictPass,
		Reason: "No high-impact events within lockout window",
	}
}

// MR-REJECT-002: No entries during Asian session for non-Asian pairs.
func checkSessionRestriction(ta *models.TASymbolResult) models.GuardCheckResult {
	if Is247Market(ta.Symbol) {
		return models.GuardCheckResult{
			Rule: constants.RuleSessionRestriction, Verdict: constants.VerdictPass,
			Reason: "Session restrictions do not apply to 24/7 markets",
		}
	}

	now := time.Now().UTC()
	hour := now.Hour()
	isAsian := hour >= 0 && hour < 7

	if !isAsian {
		return models.GuardCheckResult{
			Rule: constants.RuleSessionRestriction, Verdict: constants.VerdictPass,
			Reason: fmt.Sprintf("Current hour %d UTC is outside Asian session", hour),
		}
	}

	symbol := strings.ToUpper(ta.Symbol)
	if strings.Contains(symbol, "JPY") || strings.Contains(symbol, "AUD") || strings.Contains(symbol, "NZD") {
		return models.GuardCheckResult{
			Rule: constants.RuleSessionRestriction, Verdict: constants.VerdictPass,
			Reason: fmt.Sprintf("%s is active during Asian session", symbol),
		}
	}

	return models.GuardCheckResult{
		Rule:     constants.RuleSessionRestriction,
		Verdict:  constants.VerdictReject,
		Reason:   fmt.Sprintf("Asian session restriction: %s should not be traded 00:00-07:00 UTC", symbol),
		Metadata: map[string]interface{}{"hour_utc": hour, "symbol": symbol},
	}
}

// MR-REJECT-006: Counter-trend without CHoCH = NO SETUP.
func checkCounterTrend(processor *models.ProcessorOutput, ta *models.TASymbolResult) models.GuardCheckResult {
	if processor == nil || !processor.TradeValid {
		return models.GuardCheckResult{
			Rule: constants.RuleCounterTrendNoChoch, Verdict: constants.VerdictPass,
			Reason: "Trade not valid, guard not applicable",
		}
	}

	trend := ta.OverallTrend
	if trend == "" {
		trend = "NEUTRAL"
	}
	direction := strings.ToUpper(processor.Direction)

	isCounter := (trend == "BULLISH" && (direction == "SHORT" || direction == "BEARISH" || direction == "SELL")) ||
		(trend == "BEARISH" && (direction == "LONG" || direction == "BULLISH" || direction == "BUY"))

	if !isCounter {
		return models.GuardCheckResult{
			Rule: constants.RuleCounterTrendNoChoch, Verdict: constants.VerdictPass,
			Reason: "Trade aligns with overall multi-TF trend",
		}
	}

	totalChoch := 0
	for tf, snapshot := range ta.Snapshots {
		chochEvents, ok := snapshot["choch_events"]
		if !ok || chochEvents == nil {
			continue
		}

		switch v := chochEvents.(type) {
		case map[string]interface{}:
			if count, ok := v["count"]; ok {
				switch c := count.(type) {
				case float64:
					totalChoch += int(c)
				case int:
					totalChoch += c
				case int64:
					totalChoch += int(c)
				}
			}
		case float64:
			totalChoch += int(v)
		case int:
			totalChoch += v
		case int64:
			totalChoch += int(v)
		case []interface{}:
			totalChoch += len(v)
		default:
			// Unknown format: log for investigation but do not reject.
			// Silently returning 0 here could incorrectly reject valid
			// counter-trend trades with CHoCH confirmation.
			log := observability.Logger("guard_evaluator")
			log.Warn().
				Str("timeframe", tf).
				Str("symbol", ta.Symbol).
				Str("type", fmt.Sprintf("%T", chochEvents)).
				Msg("choch_events_unexpected_format")
		}
	}

	if totalChoch > 0 {
		return models.GuardCheckResult{
			Rule:     constants.RuleCounterTrendNoChoch,
			Verdict:  constants.VerdictWarn,
			Reason:   "Counter-trend trade with CHoCH detected across timeframes - proceed with caution",
			Metadata: map[string]interface{}{"total_choch_count": totalChoch},
		}
	}

	return models.GuardCheckResult{
		Rule:     constants.RuleCounterTrendNoChoch,
		Verdict:  constants.VerdictReject,
		Reason:   "Counter-trend trade without any CHoCH across timeframes - rejected per MR-REJECT-006",
		Metadata: map[string]interface{}{"trend": trend, "direction": direction, "htf_timeframes": ta.HTFTimeframes},
	}
}

// MR-REJECT-008: No new entries close to market close on Friday.
func checkWeekendGapRisk(ta *models.TASymbolResult) models.GuardCheckResult {
	if Is247Market(ta.Symbol) {
		return models.GuardCheckResult{
			Rule: constants.RuleWeekendGapRisk, Verdict: constants.VerdictPass,
			Reason: "Weekend gap risk does not apply to 24/7 markets",
		}
	}

	now := time.Now().UTC()

	if now.Weekday() == time.Friday && now.Hour() >= 20 {
		return models.GuardCheckResult{
			Rule: constants.RuleWeekendGapRisk, Verdict: constants.VerdictReject,
			Reason:   "Friday after 20:00 UTC - weekend gap risk",
			Metadata: map[string]interface{}{"day": "Friday", "hour": now.Hour()},
		}
	}

	if now.Weekday() == time.Saturday || now.Weekday() == time.Sunday {
		return models.GuardCheckResult{
			Rule: constants.RuleWeekendGapRisk, Verdict: constants.VerdictReject,
			Reason:   "Weekend - market closed",
			Metadata: map[string]interface{}{"day": now.Weekday().String()},
		}
	}

	return models.GuardCheckResult{
		Rule: constants.RuleWeekendGapRisk, Verdict: constants.VerdictPass,
		Reason: "Not in weekend gap risk window",
	}
}

// MR-REJECT-009: Warn during known low-liquidity hours.
func checkLowLiquidityHours(ta *models.TASymbolResult) models.GuardCheckResult {
	if Is247Market(ta.Symbol) {
		return models.GuardCheckResult{
			Rule: constants.RuleLowLiquidityHours, Verdict: constants.VerdictPass,
			Reason: "Low liquidity hours do not apply to 24/7 markets",
		}
	}

	now := time.Now().UTC()
	hour := now.Hour()

	if hour >= 21 || hour < 1 {
		return models.GuardCheckResult{
			Rule:     constants.RuleLowLiquidityHours,
			Verdict:  constants.VerdictWarn,
			Reason:   fmt.Sprintf("Low liquidity period: %d:00 UTC", hour),
			Metadata: map[string]interface{}{"hour_utc": hour},
		}
	}

	return models.GuardCheckResult{
		Rule: constants.RuleLowLiquidityHours, Verdict: constants.VerdictPass,
		Reason: fmt.Sprintf("Normal liquidity hours: %d:00 UTC", hour),
	}
}
